### PR TITLE
Refactor the entire storage system and the section editor

### LINKED ISSUE #471: Use Stylus settings in FF private mode
I lose my stylus settings pretty often: In private mode, they are always on default instead of inheriting the ones from normal mode. This should not be.

As soon as you view the Stylus settings in private mode and change something to your needs, close the private window (=delete cookies/website offline data), you also lose settings in normal mode. But when going back to private mode, they are on default again... This practically means you can only have default settings in private mode.

FF 61.0.2
Stylus 1.4.13

### LINKED ISSUE #496: Switch to usercss-meta?
https://github.com/openstyles/stylus/issues/486#issuecomment-417368790

I read through `js/usercss.js` and compared them with `usercss-meta`:

1. In `usercss.js`, we tried to keep `file://` URL in `@updateURL` at [L24](https://github.com/openstyles/stylus/blame/b90f7bfce51c308b7965f8f44d7b066227e8f24a/js/usercss.js#L24-L26), but it doesn't work because the parser would throw on non-http(s) protocols.
2. In `usercss-meta`, parsing metadata without `@name` would throw, which may affect default template saving. This is configurable.
3. `usercss-meta` doesn't expose `validateVar`, which is required to implement `assignVars`. Validation is built into the regular expression. They should probably be pulled out while adding `number` and `range`.
4. `usercss.js` uses `/@(\w+)[ \t\xA0]*/mg` to match metadata key. `usercss-meta` uses `/@(\w+)[^\S\r\n]*/mg`

### LINKED ISSUE #522: Turn all styles off doesn't work on github.com after refresh
* **Browser**:Firefox 62
* **Operating System**:Windows 10
* **Screenshot**:
screenshot

"Turn all styles off" doesn't work after refresh. Unchecking and re-checking makes it working, but it doesn't work again after another refresh.

### LINKED ISSUE #525: "turn all styles off" toggle doesn't work on many pages
* **Browser**: Firefox (Quantum) 64-bit 62.0.2
* **Operating System**: Fedora 28 Twenty Eight (x86-64), Cinnamon 3.8.9
[* **Screenshot**](https://i.imgur.com/7g4HtBD.png)
(Screenshot couldn't upload from here for some reason)

I have the stylesheet "Global dark style" installed for use at nighttime. I also have an extension called NoScript installed, and it is used to disable javascript on websites unless explicitly allowed for that domain.

I have "turn all styles off" enabled, but when opening some pages, it seems they need javascript allowed on the page in order to *disable* the stylesheet. I can easily reproduce this by opening a site to which I've never been before, as seen above with yahoo.com - but it also happens sometimes on pages that I have been to before.

It happens more often on sites whose addresses I have typed into the address bar. If I follow a link to [this site](https://apnews.com/e0ce67492cd843659d09b56c93cfeb35), the problem does not occur, but if I then click on the URL in the address bar and hit enter, the stylesheet is enabled.

I can remedy the problem by toggling the "turn all styles off" checkbox back and forth, but then a refresh of the page reverts to the stylsheet-enabled dark state again.

I wouldn't mind hearing that the extension doesn't work on pages where I've got another extension messing around and blocking javascript. But it's odd to me that Stylus works fine on these pages, and *disabling* Stylus is the function that is broken. Is there any way to work around this?

edit: after posting this issue, I reloaded this page and, ironically enough, the stylesheet was enabled. NoScript is fully disabled on this site (that is, all domains here are whitelisted) so I'm not sure what the issue is anymore. With NoScript disabled, I cannot replicate the issue, so it still seems to be a conflict between the two extensions.

edit2: scratch that, I just had the issue again, with NoScript disabled, after clicking on my own profile page here on Github. That leads me to believe it is not related to NoScript at all, although it is odd that the other websites where I was able to guarantee replication of the issue are no longer having the problem.

### PR DESCRIPTION
**Pull request by eight04**
Fixes #496.
Fixes #471.
Fixes #525.
Fixes #522.

This branch already adopted #509 and #510.

It should be relatively easy to implement #360 based on the new system.

### CODE REVIEW

**Comment by eight04** [comment]
Since the preference system becomes asynchronous, all code depending on the prefs will be affected. It would be nice if someone who is not using the default settings can test this branch.

**Comment by Mottie** [comment]
Should I ignore #509 & #510 then? Can we close them?

**Comment by eight04** [comment]
Yes. This branch is based on them and includes more bugfixes.

**Comment by Mottie** [comment]
While importing an exported set of userstyles, I see the following error:

```
background/style-manager.js:392 
Uncaught (in promise) TypeError: style.exclusions.some is not a function
  at urlMatchStyle (background/style-manager.js:392)
  at getAppliedCode (background/style-manager.js:361)
  at buildCache (background/style-manager.js:348)
  at getSectionsByUrl (background/style-manager.js:332)
  at preparing.then (background/style-manager.js:197)
```

**Comment by eight04** [comment]
I think your backup contains a style that has been modified by the original implementation of #360. #360 sets `style.exclusions` to an object but this branch assumes that `style.exclusions` is an Array.

**Comment by eight04** [comment]
You can fix it by deleting the `exclusions` property manually in the JSON file before importing.

**Comment by eight04** [comment]
Fixed in https://github.com/openstyles/stylus/pull/518/commits/0b4d12a71f343bf174fc8d4ec9613c4623d7b4b3.

**Comment by Mottie** [comment]
While trying to manually update GitHub-Dark UserCSS via the update button, I saw this error in the console:

> Uncaught (in promise) TypeError: Cannot read property 'usercssData' of undefined
    at Object.checkStyle (background/update.js:93)
    at onRuntimeMessage (background/background.js:405)
    at withPromiseError (msg.js:273)
    at executeCallbacks (msg.js:189)
    at doSend (msg.js:83)
    at withPromiseError (msg.js:273)
    at Object.sendBg (msg.js:73)
    at Proxy.args (msg.js:324)
    at checkUpdate (updater-ui.js:101)
    at HTMLAnchorElement.check (manage.js:419)

**Comment by Mottie** [comment]
Also, when clicking on the "Check all styles for updates", I see a message "No updates found" yet numerous styles are showing a red update icon indicating an update is available.

**Comment by eight04** [comment]
The red icon indicates that there is an update error. It should correctly check the update after https://github.com/openstyles/stylus/pull/518/commits/e8b64d11a3c33628e04f277918c98738db816e16.

**Comment by eight04** [comment]
Fixed in https://github.com/openstyles/stylus/pull/518/commits/cdc922dc9b7271e8854e2b1d8f098e992923cce4

**Comment by Mottie** [comment]
I noticed that if I'm editing a userstyle (`GitHub-Dark.user.css` in this case) with custom styles applied (background image & non-default base color) and "live preview" checked, any changes to the css will reload the userstyle without custom settings applied. Once the changes are saved, the custom settings are again applied and all is good.

stylus-issue

**Comment by eight04** [comment]
Fixed in https://github.com/openstyles/stylus/pull/518/commits/b9cc8b8ca5a68a8b2e5d9d1cd6a872002fb1c722

**Comment by Mottie** [comment]
Seeing this error when I open the popup and use "Find styles" with the inline checkbox checked:

> Uncaught (in promise) TypeError: styles.value(...) is not a function or its return value is not iterable
    at styleExists (background/style-manager.js:123)
    at preparing.then (background/style-manager.js:200)
styleExists @ background/style-manager.js:123
preparing.then @ background/style-manager.js:200
Promise.then (async)
processNextResult @ search-results.js:292
process @ search-results.js:254
Promise.then (async)
load @ search-results.js:236
onclick @ search-results.js:90

**Comment by eight04** [comment]
Fixed in https://github.com/openstyles/stylus/pull/518/commits/0c7e641ea973207d041425681c8e6cfc9de6bf6d

File: background/usercss-helper.js [hunk 1]
```diff
@@ -40,89 +42,120 @@
...
-  function parse(style, allowErrors = false) {
-    // restore if stripped by getStyleWithNoCode
-    if (typeof style.sourceCode !== 'string') {
-      style.sourceCode = cachedStyles.byId.get(style.id).sourceCode;
+    function doBuild(style, findDup) {
+      if (vars || assignVars) {
+        const getOld = vars ? Promise.resolve({usercssData: {vars}}) : findDup;
+        return getOld
+          .then(oldStyle => usercss.assignVars(style, oldStyle))
+          .then(() => usercss.buildCode(style));
```
[the remark this answers is not in the document]
Comment (eight04) [reply background/usercss-helper.js]: No. `usercss.assignVars` has no return value.

File: background/db.js [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+            } catch (err) {
+              reject(err);
+            }
+          }
+        },
+        onerror(event) {
+          console.warn(event.target.error || event.target.errorCode);
+          reject(event);
+        },
+        onupgradeneeded(event) {
```
[the remark this answers is not in the document]
Comment (eight04) [reply background/db.js]: I think this is a event handler:
https://developer.mozilla.org/en-US/docs/Web/API/IDBOpenDBRequest/onupgradeneeded

**Comment by eight04** [comment]
> The `tools/update-codemirror-themes.js` file gathers the list of CodeMirror themes and used to update the `edit/codemirror-editing-hooks.js` file, but that's been removed, so now it needs to target the `edit/edit.js` file:

Maybe it would be better if we make the tool output to a separated file e.g. `edit/codemirror-themes.js`?

File: edit/codemirror-factory.js [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        wrapper.classList.toggle('CodeMirror-active', wrapper.contains(document.activeElement));
+      });
+    });
+    cm.on('focus', () => {
+      rerouteHotkeys(false);
+      wrapper.classList.add('CodeMirror-active');
+      cm.lastActive = Date.now();
+    });
+    editors.add(cm);
+    cm.distroy = () => editors.delete(cm);
```
[the remark this answers is not in the document]
Comment (eight04) [reply edit/codemirror-factory.js]: Good catch. This line is not used anymore. `cmFactory.destroy(cm)` should be used instead.

File: vendor/usercss-meta/README.md [hunk 4]
```diff
@@ -0,0 +1,5 @@
+## usercss-meta v0.8.1
```
Comment (Mottie) [inline vendor/usercss-meta/README.md]: Update to v0.8.2?

File: manifest.json [hunk 5]
```diff
@@ -110,5 +114,11 @@
   "options_ui": {
     "page": "options.html",
     "chrome_style": false
+  },
+  "applications": {
+    "gecko": {
```
[the remark this answers is not in the document]
Comment (eight04) [reply manifest.json]: The `applications` key is required to make side-loaded extensions be able to use `storage.sync` API in Firefox:
https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/WebExtensions_and_the_Add-on_ID#When_do_you_need_an_add-on_ID

File: vendor/usercss-meta/README.md (hunk 4 above)
Comment (eight04) [reply vendor/usercss-meta/README.md]: I think it is better to create another PR to update libraries after this branch is merged.

**Comment by polyzen** [comment]
Bumparoni

**Comment by Mottie** [comment]
Soon grasshopper, soon.

**Pull request merged**